src/analyses/primitive_clustering.py

```diff
@@ -0,0 +1,199 @@
+    matrix, and the second element is a new labels array that names
+    the rows and columns of this new matrix
+    """
+    new_labels = deepcopy(labels_arr)
+    i = 0
+    while i < len(arr):
+        j = 0
+        accepted = False
+        while not accepted and j < len(arr[i]):
+            if arr[i][j] < cutoff and i != j:
+                accepted = True
+                break
+            j += 1
+
+        if not accepted:
+            arr = np.delete(arr, i, 0)
+            arr = np.delete(arr, i, 1)
+            del new_labels[i]
+            i -= 1
+
+        i += 1
+
+    return arr, new_labels
+
+
+class PrimitiveClusteringAnalysis(Analysis):
+    """
+    An analysis that clusters primitives by their output difference.
+    For each output difference metric, a dendrogram is saved showing
+    the clustering of primitives.
+    """
+
+    required_aggregations = [PrimitivePairComparisonAggregation]
+    CLIP_DISTANCE = 0.05
+
+    def save_dendrogram(self, linkage, labels, metric_label, *, y_cutoff):
+        print(f'Saving dendrogram for metric {metric_label}...')
+
+        plt.figure(figsize=(25, 10))
+        plt.title('Hierarchical Clustering - '+metric_label)
+        plt.xlabel('Primitive Python Path')
+        plt.ylabel('Distance')
+
+        dendrogram(linkage, labels=labels, color_threshold=0, leaf_rotation=50.0)
+
+        axes = plt.gca()
+        axes.set_ylim([0,y_cutoff])
+        plt.setp(axes.get_xticklabels(), ha="right")
+
+        if not os.path.isdir(DataDir.ANALYSIS.value):
+            os.makedirs(DataDir.ANALYSIS.value)
+        if not os.path.isdir(os.path.join(DataDir.ANALYSIS.value, self.__class__.__name__)):
+            os.makedirs(os.path.join(DataDir.ANALYSIS.value, self.__class__.__name__))
+
+        plt.savefig(
+            os.path.join(DataDir.ANALYSIS.value, self.__class__.__name__, metric_label+'.png'),
+            bbox_inches='tight'
+        )
+
+    def run(self, entity_maps: Dict[str, dict], verbose: bool, refresh: bool,
+            aggregations: Dict[str, Any]=None):
+
+        ppcm = aggregations['PrimitivePairComparisonAggregation']['ppcm']
+        prim_id_to_paths = aggregations['PrimitivePairComparisonAggregation']['prim_id_to_paths']
+
+        # Here we construct prim_indexes to map each primitive ID
+        # to what will be its index in the distance matrix.
+        #
+        # The final assignment is deliberately indented to be after the
+        # last iteration of the loop; the keys of the PPCM are ordered by
+        # the primitives' IDs, and the IDs in each pair are themselves
+        # ordered. This means that the primitive ID that is last
+        # alphanumerically will be the only one never to appear as the
+        # first primitive ID in a pair, but it will be the second ID in
+        # the final pair. So we know this ID will be bound to prim2 after
+        # the final iteration, and we add it to prim_indexes at that point.
```

Comment:
I'd love to ask you about what's happening in this paragraph @JeremyRees when I see you if that's ok.